### PR TITLE
modules/py_ml: ndarrays only need to match the input size if 1D.

### PR DESCRIPTION
**Pull request by kwagyeman**
1D ndarrays only need to match the input size. Their shape doesn't have to match.

### CODE REVIEW

**Comment by iabdalkader** [comment]
Why don't we just skip the batch size for everything ?

**Comment by iabdalkader** [comment]
Meaning when loading input/output shapes, skip the first dimension ?

**Comment by kwagyeman** [comment]
There's no way to know it's the batch size since it could be missing or present.

**Comment by iabdalkader** [comment]
As far as I know, models always have a batch size, even if not specified. Can you find/create a model without a batch size ?

**Comment by kwagyeman** [comment]
Will do a search again. When I previously did this, the answer was inconclusive. Will spend more time on it.

**Comment by kwagyeman** [comment]
Okay, I spent some more time reading up about this.

The batch size is always the first dimension for any model. When creating a model, you can specify it or leave it unspecified and have the library choose it. It defines how many samples are processed at once... and the point is to leverage your GPU's parallel nature so training goes faster. However, the larger it is, the larger the GPU memory resources need to be. Training updates are only applied per Epoch, so the batch size doesn't affect the training results.

After training is finished, the batch size for inference is 1, as you want the model to work on one item at a time. If It were two or more, this would mean you'd be asking the model to process two or more pieces of data and run the same model on them in parallel. This makes sense for GPUs with lots of resources. For us, I don't see a need for a batch size larger than 1 as you can get the same result by running the the model serially over many pieces of data.

**TLDR; we should assume the first dimension is the batch size and it should be compared to be 1 as it always should be 1.** Note that we definitely could handle a model where it's larger than 1... we'd effectively just be leaving the other batch-size inputs as garbage and ignoring their outputs. However, the user should reduce their model to using a batch size of 1 to run onboard as running a model with a larger batch size would just result in a lot of extra wasted computation if only feed a single batch even though it was processing multiple batches.

**Comment by iabdalkader** [comment]
> Thoughts? Will update this PR and #2258 afterwards.

Yeah we should remove the batch size. It's just like you said we can't waste a whole ndarray dimension, they're too expensive. The code in the backend should just skip the first dimension, start looping from the second one. I'll send it later.

As for scripts/API compatibility, I thought about this, and you'll always be safe if you use this: `ndarray.reshape(model.input_shape)`. Image are handled different anyway, and batch is ignored. 

One concern though, in general, does reshape create a deep copy ? And more specifically, if the dimensions of input and target are the same ?

> Note that we definitely could handle a model where it's larger than 1...

I don't think we should even bother with this, as it will be the same as calling the model in a loop from the top script. That said, we can store the input/output batch size in the model object *in the future* if it's needed.

**Comment by kwagyeman** [comment]
Reshape creates deep copies if the array is sparse. Otherwise, it just makes a memory view of the array. This allows the original array to be accessed in one way and the view object to access it in a different way.

**Comment by iabdalkader** [comment]
> Otherwise, it just makes a memory view of the array. 

Great, then we should always recommend a reshape using the model's input shape.
